File: hs/Rave.hs
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE ParallelListComp #-}
{-# LANGUAGE ApplicativeDo #-}
module Rave (benchmark, summary) where

import           Control.Monad.Compat

import           Data.Key (forWithKey_)
import           Data.List (genericLength)
import Data.Foldable (sequenceA_)
import           Graphics.Blank
import qualified Graphics.Blank.Style as S
import           Graphics.Blank.Style (CanvasColor)

import           Prelude.Compat

import           System.Random

import           Utils

benchmark :: CanvasBenchmark
benchmark ctx = do
    let w = width ctx
        h = height ctx
        dy = h / fromIntegral numGradients
        ys = [0, dy .. h-dy]
    rgbsList <- replicateM numGradients . replicateM numColors $
        S.rgb <$> randomIO <*> randomIO <*> randomIO
    send' ctx $ sequenceA_ [ drawGradient (0, y, w, dy) rgbs
                           | y <- ys
                           | rgbs <- rgbsList
                           ]

summary :: String
summary = "Rave"

numGradients, numColors :: Int
numGradients = 100
numColors    = 6

drawGradient :: CanvasColor c => Path -> [c] -> Canvas ()
drawGradient (gx0, gy0, gx1, gy1) cs = do
    beginPath();
    rect(gx0, gy0, gx1, gy1);
    grd <- createLinearGradient(gx0, gy0, gx1, gy0+gy1);
    let cMaxIndex = genericLength cs - 1
    forWithKey_ cs $ \i c ->
        grd # S.addColorStop (fromIntegral i/cMaxIndex, c);
    S.fillStyle(grd);
    fill();
    closePath();

File: js/Rave.js
function Rave(num, ctx) {
    var numGradients = 10;
    var numColors = 6;
    var cols = [];
    var w = window.innerWidth;
    var h = window.innerHeight;
    var dy = Math.floor(h / numGradients);
    
    var y = 0;
    for (i = 0; i < numGradients; i++) {
        for (j = 0; j < numColors; j++) {
            c1 = Math.floor(Math.random() * 255);
            c2 = Math.floor(Math.random() * 255);
            c3 = Math.floor(Math.random() * 255);
            cols[j] = String("rgb(" + c1 + "," + c2 + "," + c3 + ")");
        }
        
        ctx.beginPath()
        ctx.rect(0, y, w, dy);
        var grd = ctx.createLinearGradient(0, y, w, y + dy);
        
        var maxIndex = numColors - 1;
        for (k = 0; k < numColors; k++) {
                grd.addColorStop(k / maxIndex, cols[k]);
        }

        ctx.fillStyle = grd;
        ctx.fill();
        ctx.closePath();

        y += dy;
    }
}
